### DISCUSSION TITLE
Migration of SlickGrid to TypeScript

### DISCUSSION THREAD

**Discussion by Danielku15** [6pac/SlickGrid#577] [Ideas]
I am following up here the discussion of #572, I think it is worth having an own dedicated issue for that 😉  @6pac @ghiscoding 

It would be worth a consideration to migration SlickGrid to a typescript codebase and try to modernize the codebase by eliminating some dependencies of which some are maybe even unmaintained. 

Some benefits on that: 

- A more modern tech-stack attracts more developers to contribute. Getting new devs bootstrapped and accomodated with the SlickGrid development environment is eaiser this way. 
- Providing libraries for other projects becomes easier (NPM, TypeScript definitions,..)
- Dependency management gets more visible
- Removing jQuery makes it more attractive for new frameworks like Angular, React etc. which usually avoid jQuery. 
- Test infrastructure, documentation etc. is easier to maintain.
- Code quality overall can be improved (TSDoc, TypeSafety, OOP development style, module system compatibility,..)

If there is a strategy in this regards, I could try to prepare a basic infrastructure based on my own project which provides bundling, CI/CD scripts, etc. Or you just peek into my structure as a first idea: 

It might have a bit more than SlickGrid will need. But especially setting up rollup, typescript, karma, incremental compilation etc. took me a bit, but now it's really smooth (to me at least 😁 ). 

- https://github.com/CoderLine/alphaTab/
- https://alphatab.net/docs/contributing

**Comment by Danielku15** [comment]
@ghiscoding Only Repository Maintainers can do that 😉 Feel free to move it. 
![image](https://user-images.githubusercontent.com/674916/107660344-71786600-6c88-11eb-9cf8-319a38066f16.png)

**Reply by ghiscoding** [reply]
Right.. since I had the button I could do it lol 🤦‍♂️ 
I'll hide your last comment in a bit to cleanup the discussion. Thanks

**Comment by ghiscoding** [comment]
Feel free to start a new branch for this work if you want, we could cooperate on that same branch. 

@6pac would you be ok with that kind of approach? 
I did mention in the other post that we can target multiple bundle (ES5, ESM, CommonJS, UMD) to get it working in every environment (old and new), so that shouldn't be a worry to have it all working, the only difference might be how you import the file that might (probably would) change. How do you load them on your side anyway? Just script tag or..?

**Reply by Danielku15** [reply]
I made a very quick PoC with copy-pasting most of my configs over: https://github.com/Danielku15/SlickGrid/tree/feat/typescript

It might still contain some stuff which SlickGrid will not need but compilation and watch works already. 
I ported already the slick.core but to really test and compare it I will also need to port over the other core parts. Then maybe also some reorganization and cleanup can be done next. Still a bit to go for actually giving it a test. 

So it is a (really really) early sneek-peek.

**Reply by ghiscoding** [reply]
Wow that was fast... things to note, you can't delete every external lib since some were actually customized for them to work for SlickGrid, one such customization is in `jquery.event.drag-2.3.0.js` and without it you won't get the drag/drop of column unless we go with another external lib (for example, there's another fork [slickgrid-es6](https://github.com/DimitarChristoff/slickgrid-es6) which replaced it with `interact.js`)), there are so many pieces in this big puzzle lol

**Reply by 6pac** [reply]
I've been thinking about a breaking release 3.0 for a while, but as @ghiscoding has mentioned, eliminating the dependencies is a *huge* undertaking. We are probably talking about months of work.
Major chunks of work:
- eliminate jQueryUI
- eliminate jQuery custom drop and drag libraries (there have been two failed attempts to do this already)
- eliminate jQuery
- convert to typescipt
- create 'gridcanvas' object to encapsulate grid splitting and locking in two dimensions (possible start and end Horiz and start and end Vertical, ie. nine separate sub-grids) - the idea being that the main code just has to deal with a single grid
- add dataprovider (I have this code already in my private branch)

**Reply by Danielku15** [reply]
> you can't delete every external lib since some were actually customized for them to work for SlickGrid

@ghiscoding but it should be the goal. 😉 A library like slickgrid should never require to modify it's dependencies, that's quite a bad practice. Dependencies should be consumed from NPM and should not be bundled with the library. Others will want to install SlickGrid, and get the official flavors of the other libraries via NPM too. They might in fact even use the dependency also directly in their project and won't want the SlickGrid specific adjustments. 

But thanks for pointing this out, I hope there are not many customizations and they are somehow visible in the history or by diffing. 

> eliminating the dependencies is a huge undertaking. 

@6pac  Some dependencies can still remain if they are reasonable, also it can be a step-by-step process. What definitly should be tackled early is replacing dependencies which are really outdated and not maintained anymore. Just some examples: 

- jQuery UI is dead (in terms of being a proper library nowadays). The last release was 4 years go. This library might imply risks security wise, and forces you potentially to use stone-age old jQuery which also imply security risks to your application (e.g. XSS). But anyhow only the pre-shipped editors require this. I bet these can also be replaced easily with plain HTML flavors. 
- jQuery drag&drop: It does not seem to be such a huge library, potentially the required bits can be ported into SlickGrid itself. 
- jQuery: This will not be easy, most likely it will remain for a bit, but upgrade to a newer version would be beneficial. 
- Conversion to typescript: this is indeed a tricky task but is also possible step-by-step. Type annotations, null-safety, usage of new language featrures etc. can be added step by step after the initial port. TypeScript allows quite some configs to not enforce everything from the start. 

About the other things you know best what additional changes beside a port make sense. 😉

**Reply by Danielku15** [reply]
Porting to TypeScript is done, at least to the state where it compiles. But there is still a bit to do before it would work: 

- There are quite some places I need to double check regarding the usage of `this`. Typescript is way more strict there than JavaScript and I need to ensure that with the passing on of methods as event listeners and also with the usage of arrow functions the `this` context is everywhere correct. 
- Due to the bundling with rollup the previous splitting cannot be achieved so easily. Half of the files are more "example code" than proper production code. Might be historically like that and could be questioned for a 3.0. Some exports are even overlapping. I still need to check how we can achieve the same file-structure like before. Currently rollup bundles the "core" part directly into the other modules if they are included. 
- I found already various places which might indicate bugs or dead code. These will need to be cleaned (usually visible as `as any` casts to satisfy TypeScript)
- Strong types should be introduced everywhere (currently most of the code is untyped)

Once we reached the state the SlickGrid compiles and works as before, the cleanup and modernization could start. But the initial goal should be to just have the same library like before (from a user perspective).

At this point I also fear I will have to stop with this proof-of-concept but I hope it gave you an insight/outlook how things could go.

**Reply by ghiscoding** [reply]
- For the `this` issue, there's the new `globalThis` that I never had a chance to try that might fix some of the problems you encountered
- On the strong Typings subject, that part will be easy, I already have all the Types in my libs which took me quite some time to write... so you can just copy them from my [/interfaces](https://github.com/ghiscoding/slickgrid-universal/tree/master/packages/common/src/interfaces) folder, there's also the [/enums](https://github.com/ghiscoding/slickgrid-universal/tree/master/packages/common/src/enums) that goes with it. Of course there are some extra interfaces/properties which are specific to my SlickGrid wrapper libs but at least there's no need to start from scractch.
- If you get it all to build, the best is to run the Cypress tests, they will help for sure to make sure we aren't breaking anything. There's about 100 E2E tests in this fork but I also have over 400 of them in my libs.
- I agree with your comment _"Once we reached the state the SlickGrid compiles and works as before"_

**Reply by 6pac** [reply]
Agree on all of these, other than that I'm dreading getting rid of the drag/drop libs. It may have evolved, but last time I looked browser support and third party libraries were all over the place with feature support and feature implementation.  
Also I generally follow the practice of putting ```var self = this;``` at the start of a declaration where ```this``` needs to be used consistently. I think policies around this kind of language construct can avoid a lot of the issues, which I agree can be very subtle and hard to diagnose.  
I'm currently searching for some subcontractors or employees, it could possibly be something a junior developer could churn through. Would actually also be an awesome learning exercise.

**Reply by 6pac** [reply]
can't help wondering if the jquery elimination could be largely accomplished with some advanced pattern recognition engine, like a successor to GREP.

**Reply by ghiscoding** [reply]
it's not that easy to replace jQuery, it's better to do it by hand because querying with jQuery will always return a jQuery object for chaining purposes (even when the element doesn't exist) while in plain JS it will return `undefined`, so there's more work to make sure you're not trying to read an object property from `undefined` which won't work, but with jQuery it wasn't much of a problem... oh and the chaining is also a missing piece since I just talked about it. Another thing that jQuery provides is event namespace, which I use in the auto-resizer plugin [here](https://github.com/6pac/SlickGrid/blob/master/plugins/slick.resizer.js#L121) `$(window).on('resize.grid.' + _gridUid`, I don't think there's any equivalent in plain JS but I don't think we have too many of these lines either. 

The best way to start is to start from the outside in, rewrite the small piece first (plugins, controls, ...) and then go into the core one as the final piece. If @Danielku15 code is all working, I can start with rewriting some of them, but I don't want to start if it's not ready to start

**Reply by Danielku15** [reply]
For replacing such libraries like jQuery in a big fashion, I made in the past good experience with just developing an own library with the same operations, and then use IDE "inline method" or refactoring features. This way you can advance step-by-step on replacing certain method calls while still having some jQuery parts within your "wrapper". 

For the `this` scoping with lambdas using directly the TypeScript compiler API would be my choice of analysis engine. Like a custom linter for this purpose. Most places are already covered by TypeScript itself because it knows where `this` points to and what members it has. But places like the following one are tricky because when the method is called, `this` will not point to the Grid instance anymore within the method: 

https://github.com/Danielku15/SlickGrid/blob/8ab341690b64527e90463cd24c86e3ff4d552cdd/src/slick/Grid.ts#L720

This would need to become:
```ts
this.selectionModel.onSelectedRangesChanged.subscribe(this.handleSelectedRangesChanged.bind(this));
// or 
this.selectionModel.onSelectedRangesChanged.subscribe((e,ranges) => this.handleSelectedRangesChanged(e, ranges));
```
Most places are easy to find because they are at the same places (jQuery.on, subscribe etc.)

**Reply by ghiscoding** [reply]
I corrected your typo, it should `.bind(this)` not `,bind(this)` and yes I use that a lot in my SlickGrid wrapper libs, I'm more concerned about what I wrote in my previous post but yes I agree that we could maybe write a wrapper that even if the DOM element is not found we would still return an object, easier to go step by step as you said.

**Comment by JacobMGEvans** [comment]
Greetings,
Not sure where the progress is on this migration, however, If you are interested we can "Raid" your repository, there would be one big PR from our Organization, for a large specific task/problem domain; in this case TS conversion.

We would need a maintainer as a point of contact, we prefer using Discord https://discord.gg/xkdkAJZsBR

Let me know if you are interested.
You can read more about what we are doing here: https://github.com/OpenSourceRaidGuild/tutorial-quest
This is also an example of how we work and the workflow: HospitalRun/hospitalrun-frontend#2516

cc: @nobrayner @mpeyper @Aprillion

**Reply by ghiscoding** [reply]
Hey there, I will speak for Ben (@6pac) and @Danielku15, I assume they will add their own comments. I'll also mine below.

I can provide my own experience as being a major contributor to this fork and @6pac can provide more details. As it is now, I think @Danielku15 was mainly showing what is possible and he had to step away because of lack of time, I'm in the same boat, I contributed a lot, and still are, to this repo but I also have my own repos to take care of (I have 3 libs which are wrapper on top of this fork, that is [Angular-Slickgrid](https://github.com/ghiscoding/Angular-Slickgrid), [Aurelia-Slickgrid](https://github.com/ghiscoding/aurelia-slickgrid) and [Slickgrid-Universal](https://github.com/ghiscoding/slickgrid-universal) which are all written in TypeScript). The main issue is just a lack of time, I would love to see this lib converted to TypeScript but that would take time which I don't personally have, on the other hand I contributed and still contribute a lot of fixes and features because we use them at my work (but again I can't justify to my manager that I would need to spend the next few days, probably weeks, on rewriting SlickGrid to TypeScript, I've already spent a lot of work time in fixes & features already).

On a side note, I'll speak for Ben (@6pac) here, Ben never used TypeScript but he knows .Net and I speak for him in believing that he'd be ok to move this to TypeScript (he wants to learn TS eventually) if it's just a matter of merging a PR and releasing a `major` (breaking change) version. Ben please reply if I misunderstood your position 😉 

So all that to say, if you do want to work on this and provide a PR, I could be a reviewer of the PR and also help in some areas (as time permits of course). I have knowledge of TypeScript and my libs, which again are all wrapper on top of SlickGrid, are all written in TypeScript, I have all necessary interfaces of SlickGrid in my own repos (a lot of them) and I can help in that regard.

in short, it's a lack of time, if you want to take it over please please please go ahead, we would all be happy if anyone (you) would like to lead the push 🚀

**Reply by Danielku15** [reply]
I can confirm from my side what @ghiscoding  wrote. I gave it a quick shot to start the migration of SlickGrid using my own library as base but due to the lack of time I am not able to fully complete such a "Raid" on reworking SlickGrid (migration, modernization, tests,..) -  even though it would be nice to do so.

Even though I am not an official maintainer but just a mini-contributor here my input on the raid: As a theme/topic/scope I would set "Migration SlickGrid to a TS codebase/ecosystem" 

Requirements: 
- Migration of codebase to typescript with project toolchain (build, test, deploy)
- Minimal or no breaking surface API changes  
- Adaption/Integration/Replacement of modified 3rd party dependencies 
- ES5 compatibility

Open for discussion: 
- Reorganizing/merging shipped bundles/partials (core, grid, dataview etc.) 

Non-Goals: 
- Elimination of jQuery/jQueryUI
- General modernization and improvements of codebase/architecture/logic.

**Reply by JacobMGEvans** [reply]
Outstanding, I will take this as a green light. If any of you or all of you can join the Discord as a point of contact for the Guild Members working on this, that would be great. Things start to move fast once the environment is set up and we finish discussing strategy.

**Reply by JacobMGEvans** [reply]
Once in the Discord ping myself or another Officer to add a Maintainer Role 😄

**Reply by 6pac** [reply]
Will join discord now.

I'd strongly suggest doing the TypeScript conversion as a single task, and then going on to dependency removal as a separate issue (I think that may end up being a lot harder than it looks for some dependencies).

Since we can convert TS to JS, I'd request if possible that the conversion be done in such a way that when the TS is converted to JS, the result is as close as possible to the original repo. That would provide a really easy baseline for us to use if we need to troubleshoot any issues.

**Reply by ghiscoding** [reply]
@6pac we don't have full control of what TypeScript will output, what we can choose is a target ES5, ES6, ES2015, .... CommonJS, AMD, You can see some examples on the [TypeScript PlayGround](https://typescript-play.js.org/?target=1#example/using-classes) though I agree that the first thing to do in a PR would be to first convert to TypeScript and get it working without removing any of the external dependencies like jQuery, jquery.event.drag-2.3.0.js and others. 

@JacobMGEvans for the TypeScript interfaces, you can take them directly from my lib, which are wrapper written in TypeScript and using this repo, you can take them from these folders [interfaces](https://github.com/ghiscoding/slickgrid-universal/tree/master/packages/common/src/interfaces) and [types](https://github.com/ghiscoding/slickgrid-universal/tree/master/packages/common/src/enums) (which includes enums & types), the starting interface would be the [slickNamespace.interface.ts](https://github.com/ghiscoding/slickgrid-universal/blob/master/packages/common/src/interfaces/slickNamespace.interface.ts). Side note, my interfaces have a lot more flags/options, for example I have Grid Export Services which SlickGrid doesn't, but it shouldn't be a problem since they're all optional. Any questions, feel free to ask, moving this lib to TypeScript would make it so much easier to import in my libs

**Reply by 6pac** [reply]
It's a bit of a double whammy for me (dinosaur that I am) - being a dotNet developer I use JS only in a web site context, so I've never used NodeJs or any of the front end frameworks. OR Typescript. You guys know of any good quality resources to quickstart issues around NodeJs?

**Reply by 6pac** [reply]
I hasten to add, I've seen a lot of trends come and go, so I tend to be a really late adopter, only when I think something is actually going to stick around for more than a few years. I have to say I am not completely convinced that TypeScript isn't the latest trend that will disappear in a few years (noting also it's developed by M$, the king of poorly conceived and short lived ideas in the tech space).  
And I've previously experienced huge problems with frameworks that have similar philosophies to React once the pages get too complex. They are great for simple stuff. So I actually actively avoid that.   
  
So don't want to dampen anyone's enthusiasm, I just feel like I should explain why I have have reservations and have been slow to take up new developments. It's not that I don't like new stuff, but I've been burnt soooo many time before.

**Reply by ghiscoding** [reply]
@6pac I have been coding all my libs with TypeScript, which as you know are wrapper on top of your repo and all I see is benefit, I catch a lot of bugs early in the process, that is also the biggest I've pushed so many fixes in your repo for the past 3 years. We also get intellisense out of the box with TypeScript and the language keeps growing in adoption every year (every framework and I really every framework, are now written in TypeScript), TypeScript is top 5 in most used languages and asked questions on Stack Overflow, so I don't see it going anywhere anytime soon. Adding Typing to a project is just making it so much more easy to catch bug early before shipping to production, can you imagine coding C# without types? Since you know .Net, I assume C#, it is very very similar (it is made by the same company) and you see it in the language, there's ton of available for example some starters on MS website [here](https://docs.microsoft.com/en-us/learn/paths/build-javascript-applications-typescript/) and even better some youtube learning video, like this guy he's really good at teaching, check this [video](https://www.youtube.com/watch?v=BwuLxPH8IDs). Going to TypeScript would also benefit in importing the lib, I just see benefit at the end of the day, don't let this scare you, it's quite easy for a .Net developer to get started with TypeScript, the only real difference is the types are on the right in TypeScript while being on the left in C# and that's about it, the rest is nearly the same (again because it's the same guys who created C# that created TypeScript) and it's built-in in VSCode. Hope that helps. 

I really suggest you to watch that youtube [video](https://www.youtube.com/watch?v=BwuLxPH8IDs) and after watching the hour video, I'm sure you'll understand why I'm pushing for the change

**Reply by 6pac** [reply]
Thanks @ghiscoding, it's not at all that I'm worried about it, I'm sure I can learn it in a flash if I want to. Strongly typed languages have been my bread and butter for two decades.  But personally it's not typing, or lack of it, that's important, it's the language design, and I think JS is extrememly well designed. JS is related more to LISP, the ancient machine learning language, than C. It's specifically designed for flexibility and the ability to 'meta' up. It can do a bunch of stuff that would be wildly difficult in C# just because of its different typing system.  I'm just not at all convinced that strong typing actually belongs in JS, although I am aware that you can type as 'Any' in TS if you don't want strong typing.

Just for the record, I'd observe that .NET was stolen more or less flagrantly from Sun (Java) - the reason it's good is specifically because it *wasn't* designed by Microsoft. and I suspect MS hired the entire team from Sun for further development - and Java was designed by some of the best developers in the world.
Not that I'm hating on MS on principle,  I have nothing against them, other than that they consistently keep dishing up deeply flawed and poorly thought through products. I've been using their products professionally since 1992, because you can't not if you're a business developer.

HOWEVER, clearly everyone wants to do this, so I'm not going to hold it back. It's time for me to embrace TypeScript.
I just feel that I have to state my position.

To clarify what I'm asking, I'm sure TypeScript will be a piece of cake, I'm more concerned about learning NodeJs, because it's all about packaging and module management rather than language per se.   
Are you aware of any good tutorials around that? I'm sure there are 1000's of resources out there, it's just that finding really good ones can take time, something I don't have a lot of.

**Reply by ghiscoding** [reply]
Just to be clear, TypeScript is a superset of JavaScript, the output of TypeScript build is JavaScript, we just need to specify which browser or EcmaScript version we want to target, in our case we'll target ES5 so it still works in legacy browser and again that would be ES5 JavaScript which is what the browser will use, TypeScript is only used when coding and running the build and it's not used in the browser, when it runs the build, it removes all the Types, Interfaces, etc... only to be left with JavaScript.

As for NodeJS, it's only used to run the TypeScript build, nothing else (or maybe not even needed if you use VSCode to build, I'm not fully sure)

**Reply by 6pac** [reply]
OK, I think I'm relatively across typescript now. It's better than I thought it would be - less prescriptive. In particular, getting intellisense will be a major bonus, it's been a long wait for that.  
I have friends who have told me that Microsoft is actually changing its culture recently - perhaps it's true!

Re NodeJs, I'm aware that it's completely unrelated to Typescript. It's just something that I need to learn at some point in order to be across modern js packaging and delivery, and it may as well be now. Don't worry about it, I'll dive in shortly.

**Comment by ghiscoding** [comment]
this is now done, so I can close this Discussion :)
